src/Draco.Compiler/Syntax/ParseTreeVisitor.cs

```diff
@@ -0,0 +1,147 @@
+    T VisitReturnExpr(Expr.Return expr);
+    T VisitLiteralExpr(Expr.Literal expr);
+    T VisitFuncCallExpr(Expr.FuncCall expr);
+    T VisitIndexExpr(Expr.Index expr);
+    T VisitVariableExpr(Expr.Variable expr);
+    T VisitMemberAccessExpr(Expr.MemberAccess expr);
+    T VisitUnaryExpr(Expr.Unary expr);
+    T VisitBinaryExpr(Expr.Binary expr);
+    T VisitGroupingExpr(Expr.Grouping expr);
+}
+
+/// <summary>
+/// Provides a base implementation of <see cref="IParseTreeVisitor{T}"/>.
+/// </summary>
+/// <typeparam name="T">The return type of the visitor.</typeparam>
+internal abstract class BaseParseTreeVisitor<T> : IParseTreeVisitor<T>
+{
+    public virtual T VisitNode(ParseTree node) => node switch
+    {
+        CompilationUnit compilationUnit => this.VisitCompilationUnit(compilationUnit),
+        Decl decl => this.VisitDecl(decl),
+        FuncParam param => this.VisitFuncParam(param),
+        FuncBody body => this.VisitFuncBody(body),
+        TypeExpr typeExpr => this.VisitTypeExpr(typeExpr),
+        TypeSpecifier specifier => this.VisitTypeSpecifier(specifier),
+        Stmt stmt => this.VisitStmt(stmt),
+        Expr expr => this.VisitExpr(expr),
+
+        _ => throw new InvalidOperationException()
+    };
+
+    public abstract T VisitCompilationUnit(CompilationUnit compilationUnit);
+
+    public virtual T VisitDecl(Decl decl) => decl switch
+    {
+        Decl.Func func => this.VisitFuncDecl(func),
+        Decl.Label label => this.VisitLabelDecl(label),
+        Decl.Variable variable => this.VisitVariableDecl(variable),
+
+        _ => throw new InvalidOperationException()
+    };
+    public abstract T VisitFuncDecl(Decl.Func decl);
+    public abstract T VisitLabelDecl(Decl.Label decl);
+    public abstract T VisitVariableDecl(Decl.Variable decl);
+
+    public abstract T VisitFuncParam(FuncParam param);
+
+    public virtual T VisitFuncBody(FuncBody body) => body switch
+    {
+        FuncBody.BlockBody block => this.VisitBlockFuncBody(block),
+        FuncBody.InlineBody inline => this.VisitInlineFuncBody(inline),
+
+        _ => throw new InvalidOperationException()
+    };
+    public abstract T VisitBlockFuncBody(FuncBody.BlockBody body);
+    public abstract T VisitInlineFuncBody(FuncBody.InlineBody body);
+
+    public virtual T VisitTypeExpr(TypeExpr typeExpr) => typeExpr switch
+    {
+        TypeExpr.Name name => this.VisitNameTypeExpr(name),
+
+        _ => throw new InvalidOperationException()
+    };
+    public abstract T VisitNameTypeExpr(TypeExpr.Name typeExpr);
+
+    public abstract T VisitTypeSpecifier(TypeSpecifier specifier);
+
+    public virtual T VisitStmt(Stmt stmt) => stmt switch
+    {
+        Stmt.Decl decl => this.VisitDeclStmt(decl),
+        Stmt.Expr expr => this.VisitExprStmt(expr),
+
+        _ => throw new InvalidOperationException()
+    };
+    public abstract T VisitDeclStmt(Stmt.Decl stmt);
+    public abstract T VisitExprStmt(Stmt.Expr stmt);
```

Comment:
Same, as well as for the other abstract methods, shouldn't they recursively visit their subelements?